### PR TITLE
Update neo4j formula

### PR DESCRIPTION
**Pull request by tristan00b**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Does your submission pass `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Have you built your formula locally prior to submission with `brew install <formula>`?

~~Failed audit due to missing `test do` block, which was not included in the original formula.~~

### CODE REVIEW

**Comment by apjanke** [comment]
Test is failing.


```
        Testing neo4j
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20160406-42628-e3npnb.sb /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Library/Homebrew -- /usr/local/Library/Homebrew/test.rb /usr/local/Library/Taps/homebrew/homebrew-core/Formula/neo4j.rb --verbose
==> /usr/local/bin/neo4j start
Unable to find any JVMs matching version "1.7".
Starting Neo4j Server...WARNING: not changing user
/usr/local/bin/neo4j: line 206: /usr/local/Cellar/neo4j/2.3.3/libexec/data/neo4j-service.pid: Operation not permitted
/usr/local/bin/neo4j: line 203: /usr/local/Cellar/neo4j/2.3.3/libexec/data/log/console.log: Operation not permitted
cat: /usr/local/Cellar/neo4j/2.3.3/libexec/data/neo4j-service.pid: No such file or directory
process []... waiting for server to be ready. Failed to start within 120 seconds.
Neo4j Server may have failed to start, please check the logs.
rm: /usr/local/Cellar/neo4j/2.3.3/libexec/data/neo4j-service.pid: No such file or directory
Error: neo4j: failed
Failed executing: /usr/local/bin/neo4j start
/usr/local/Library/Homebrew/formula.rb:1481:in `block in system'
/usr/local/Library/Homebrew/formula.rb:1418:in `open'
/usr/local/Library/Homebrew/formula.rb:1418:in `system'
/usr/local/Library/Taps/homebrew/homebrew-core/Formula/neo4j.rb:35:in `block in <class:Neo4j>'
/usr/local/Library/Homebrew/formula.rb:1327:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:36:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:1323:in `run_test'
/usr/local/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/test.rb:27:in `<main>'
```

Maybe the formula needs a Java dependency?

**Comment by tribayf** [comment]
Thanks @apjanke, I've added the Java dependency. There also seemed to be an issue with permissions. I've configured neo4j to write to `testpath`. Seems to work on my machine. Hopefully it passes now.

PS -- I'm totally new to working with Homebrew formulae (and Ruby). Any comments on style would be much appreciated!

**Comment by tribayf** [comment]
Not sure how to get this working. I think I've fixed the permissions issue, but neo4j is still failing to start.

Seems that Jenkins can't load the JVM:
`Unable to find any JVMs matching version "1.7".`

**Comment by DomT4** [comment]
Does `neo4j` need Java 1.7 _or_ anything above and including Java 1.7? The test error seems to hint that it always expects Java 1.7.

File: Formula/neo4j.rb [hunk 1]
```diff
@@ -30,4 +32,12 @@ def install
       wrapper.java.additional.4=-Dneo4j.ext.udc.source=homebrew
     EOS
   end
+
+  test do
+    ENV["NEO4J_LOG"] = testpath/"libexec/data/log/neo4j.log"
+    ENV["NEO4J_PIDFILE"] = testpath/"libexec/data/neo4j-service.pid"
+    mkpath testpath/"libexec/data/log"
+    system "#{HOMEBREW_PREFIX}/bin/neo4j", "start"
+    system "#{HOMEBREW_PREFIX}/bin/neo4j", "stop"
```
Comment (ilovezfs) [inline Formula/neo4j.rb]: `bin/"neo4j"` or `"#{bin}/neo4j"`

File: Formula/neo4j.rb [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
-    sha256 "2a20f420e94fe4189363ce8ab327c0e5e054df3fc74a0249e9e2c7fe0455a0d6"
-    version "3.0.0-M02"
+    url "http://dist.neo4j.org/neo4j-community-3.0.0-M05-unix.tar.gz"
+    sha256 "2b7878f424859de6951e86f9abb71701d8d45ab22e1157410504fe6a6bb17a95"
+    version "3.0.0-M05"
   end
 
   bottle :unneeded
 
+  depends_on :java => "1.7+"
```
Comment (mikemcquaid) [inline Formula/neo4j.rb]: Try `depends_on :java => "1.7"` as it sounds like it needs 1.7 specifically.

File: Formula/neo4j.rb (hunk 1 above)
Comment (tribayf) [inline Formula/neo4j.rb]: Awesome, thanks!

**Comment by tribayf** [comment]
@DomT4, @mikemcquaid the minimum requirement is for either Java 7 or 8, depending on what version of Neo4j, but they still recommend using the latest version of Java 8.

[Under system requirements for v2.3.3](http://neo4j.com/docs/stable/deployment-requirements.html):
> Important
> It is strongly recommended to use an up-to-date Java release as there are issues in early versions of both Java 7 and Java 8 that are known to adversely affect Neo4j.

[Under system requirements for v3.0.0-M05](http://neo4j.com/docs/3.0.0-M05/deployment-requirements.html):
> Important
> It is strongly recommended to use an up-to-date Java release as there are issues in early versions of Java 8 that are known to adversely affect Neo4j.

I looked into how Java is located by `bin/neo4j` and found the following:

For v2.3.3 via `bin/utils`:
```bash
findjava() {
  # OS specific support.  $var _must_ be set to either true or false.
  cygwin=false;
  darwin=false;
  case "`uname`" in
    CYGWIN*) cygwin=true ;;
    Darwin*) darwin=true
           if [ -z "$JAVA_VERSION" ] ; then
             JAVA_VERSION="CurrentJDK"
           else
             echo "Using Java version: $JAVA_VERSION"
           fi
           if [ -z "$JAVA_HOME" ] ; then
             JAVA_HOME=`/usr/libexec/java_home -v 1.7`
           fi
           ;;
  esac

  if [ -z "$JAVA_HOME" ] ; then
    if [ -r /etc/gentoo-release ] ; then
      JAVA_HOME=`java-config --jre-home`
    fi
  fi
[snip]
}
```

For v3.0.0-M05 directly in `bin/neo4j`
```bash
find_java_home() {
  [[ "${JAVA_HOME:-}" ]] && return

  case "${DIST_OS}" in
    "macosx")
      JAVA_HOME="$(/usr/libexec/java_home -v 1.8)"
      ;;
    "gentoo")
      JAVA_HOME="$(java-config --jre-home)"
      ;;
  esac
}

find_java_cmd() {
  [[ "${JAVA_CMD:-}" ]] && return
  detect_os
  find_java_home

  if [[ "${JAVA_HOME:-}" ]] ; then
    JAVA_CMD="${JAVA_HOME}/bin/java"
  else
    if [ "${DIST_OS}" != "macosx" ] ; then
      # Don't use default java on Darwin because it displays a misleading dialog box
      JAVA_CMD="$(which java || true)"
    fi
  fi

  if [[ ! "${JAVA_CMD:-}" ]]; then
    echo "ERROR: Unable to find Java executable."
    show_java_help
    exit 1
  fi
}
```

Does this give any hints WRT the Jenkins config? One my machine I'm able to install both versions of neo4j running Java 8 without any problems.

**Comment by mikemcquaid** [comment]
I'm out of my depth here and not really sure where to go next. You can see the Java version on the Jenkins box with http://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/247/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/config/

**Comment by apjanke** [comment]
Whatever their doco is saying, that 2.3.3 is looking specifically for Java 1.7, and 3.0.0 is looking specifically for 1.8.

For v2.3.3 via bin/utils:
```
findjava() {
  ...
           if [ -z "$JAVA_HOME" ] ; then
             JAVA_HOME=`/usr/libexec/java_home -v 1.7`
           fi
   ...
```

That `java_home -v 1.7` means "exactly version 1.7"; "1.7 or later" is `java_home -v 1.7+`. That's consistent with the `Unable to find any JVMs matching version "1.7"` error message; that's what `/usr/libexec/java_home` outputs when it can't find what it wants.

Since the `neo4j` doco for 2.3.3 says it works on Java 7 or Java 8 and even prefers Java 8, they may just accidentally be using the wrong arguments.

Could you submit this as an upstream bug report to `neo4j`? If this is really the case, then it should be a quick fix, and we can patch it in the formula to get it working now until a new release comes out.

**Comment by apjanke** [comment]
Once that part is fixed, then the test will fail due to an interaction between the sandbox and JRE's new logging, as described in https://github.com/Homebrew/brew/issues/64.

```
$ brew test --verbose neo4j
Testing neo4j
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20160410-45054-1ru5j4o.sb /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Library/Homebrew -- /usr/local/Library/Homebrew/test.rb /usr/local/Library/Taps/homebrew/homebrew-core/Formula/neo4j.rb --verbose
==> /usr/local/Cellar/neo4j/2.3.3/bin/neo4j start
Starting Neo4j Server.../private/tmp/neo4j20160410-45055-787qej/libexec/data/log/neo4j.log was missing, recreating...
WARNING: not changing user
process [45165]... waiting for server to be ready.. Failed to start within 120 seconds.
Neo4j Server may have failed to start, please check the logs.
==> Sandbox log
Apr 10 21:00:26 sandboxd[122]: java(45145) deny file-write-data /Users/janke/.oracle_jre_usage/42d8591bd1e10f3f.timestamp
Apr 10 21:00:26 sandboxd[122]: java(45147) deny file-write-data /Users/janke/.oracle_jre_usage/42d8591bd1e10f3f.timestamp
Apr 10 21:00:26 sandboxd[122]: java(45165) deny file-write-data /Users/janke/.oracle_jre_usage/42d8591bd1e10f3f.timestamp
Error: neo4j: failed
Failed executing: /usr/local/Cellar/neo4j/2.3.3/bin/neo4j start
/usr/local/Library/Homebrew/formula.rb:1481:in `block in system'
/usr/local/Library/Homebrew/formula.rb:1418:in `open'
/usr/local/Library/Homebrew/formula.rb:1418:in `system'
/usr/local/Library/Taps/homebrew/homebrew-core/Formula/neo4j.rb:44:in `block in <class:Neo4j>'
/usr/local/Library/Homebrew/formula.rb:1327:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:36:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:1323:in `run_test'
/usr/local/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/test.rb:27:in `<main>'
```

Fixing this will either take hacking system properties inside the formula, or adjusting the sandbox. I think we want to do a global fix since this will affect a lot of `java` usage.

**Comment by apjanke** [comment]
Okay, it's a further sandbox error because `neo4j`'s default configuration is set to write log files under its installation location in the cellar.

```
2016-04-10 22:16:44.450-0400 ERROR Failed to start Neo Server on port [unknown port] Bad value 'conf/neo4j-http-logging.xml' for setting 'org.neo4j.server.http.log.config': HTTP log directory [/usr/local/Cellar/neo4j/2.3.3/libexec/data/log] is not writeable.
org.neo4j.graphdb.config.InvalidSettingException: Bad value 'conf/neo4j-http-logging.xml' for setting 'org.neo4j.server.http.log.config': HTTP log directory [/usr/local/Cellar/neo4j/2.3.3/libexec/data/log] is not writeable.
```

Not sure what the right approach here is. Typically we'd relocate the logs under `$(brew --prefix)/var/log` like a typical Unix program. But Java programs tend to like to stick all their data under their main installation location; this would diverge from its documentation.

**Comment by apjanke** [comment]
I replaced with #202, which fixes the discussed problems.

Backed the test off to do a trivial `neo4j info` instead of starting the server, because the current dir structure is incompatible with the sandbox. Will address that separately.

**Comment by apjanke** [comment]
neo4j 2.3.3 is merged now.

**Comment by tribayf** [comment]
Brilliant @apjanke, thank you so much.

**Pull request closed without merging**